### PR TITLE
Update mParticle schemas

### PR DESCRIPTION
**Pull request by SyedWasiHaider**
This PR makes updates to the schema for mParticle.

I've done basic validation using `igluctl lint`.  Please let me know if I'm missing anything.

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

📝 Please visit https://github.com/snowplow/snowplow/wiki/CLA to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by SyedWasiHaider** [comment]
I signed it!

**Comment by snowplowcla** [comment]
Confirmed! @SyedWasiHaider has signed the Software Grant and Corporate Contributor License Agreement. Thanks so much

File: schemas/com.mparticle.snowplow/impression_event/jsonschema/1-0-0 [hunk 1]
```diff
@@ -0,0 +1,68 @@
+{
+	"$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+	"description": "Schema for an impression event",
+	"self": {
+		"vendor": "com.mparticle.snowplow",
+		"name": "impression_event",
+		"format": "jsonschema",
+		"version": "1-0-0"
+	},
+	"definitions": {
```
Comment (alexanderdean) [inline schemas/com.mparticle.snowplow/impression_event/jsonschema/1-0-0]: We haven't used this approach before with schemas. Does this definitely work with Snowplow - have you tested it?

File: schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0 [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+			"type": "string"
+		},
+		"launchReferral": {
+			"type": "string"
+		},
+		"successfullyClosed": {
+			"type": "boolean"
+		},
+		"jsonPayload": {
+			"type": "number"
```
Comment (alexanderdean) [inline schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0]: This sounds fishy - jsonPayload as a number?

File: schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0 [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	"self": {
+		"vendor": "com.mparticle.snowplow",
+		"name": "appstatetransition_event",
+		"format": "jsonschema",
+		"version": "1-0-0"
+	},
+	"type": "object",
+	"properties": {
+		"transitionType": {
+			"type": "string"
```
Comment (alexanderdean) [inline schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0]: No maxLengths on any strings?

File: schemas/com.mparticle.snowplow/pushregistration_event/jsonschema/1-0-0 [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	},
+	"type": "object",
+	"properties": {
+		"name": {
+			"type": "string"
+		},
+		"registrationToken": {
+			"type": "string"
+		},
+		"attributes": {}
```
Comment (alexanderdean) [inline schemas/com.mparticle.snowplow/pushregistration_event/jsonschema/1-0-0]: What does this mean?

File: schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0 (hunk 2 above)
Comment (SyedWasiHaider) [reply schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0]: Yes good catch. This should be a string.

File: schemas/com.mparticle.snowplow/pushregistration_event/jsonschema/1-0-0 (hunk 4 above)
Comment (SyedWasiHaider) [reply schemas/com.mparticle.snowplow/pushregistration_event/jsonschema/1-0-0]: This is meant to be a dictionary for any key value pairs. Seems like we had it for our AppEvent in our V1: https://github.com/snowplow/iglu-central/blob/master/schemas/com.mparticle.snowplow/app_event/jsonschema/1-0-0 so I assume this should be fine.

Comment (alexanderdean) [reply schemas/com.mparticle.snowplow/pushregistration_event/jsonschema/1-0-0]: Cool, thanks for clarifying

File: schemas/com.mparticle.snowplow/impression_event/jsonschema/1-0-0 (hunk 1 above)
Comment (SyedWasiHaider) [reply schemas/com.mparticle.snowplow/impression_event/jsonschema/1-0-0]: I was going off what I saw in the SendGrid schema: https://github.com/snowplow/iglu-central/blob/master/schemas/com.sendgrid.sauna.commands/send_email/jsonschema/1-0-0#L16

Is that not correct?

Comment (alexanderdean) [reply schemas/com.mparticle.snowplow/impression_event/jsonschema/1-0-0]: Oh nice, I hadn't seen that one! Cool, that should work then

File: schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0 (hunk 3 above)
Comment (SyedWasiHaider) [reply schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0]: Is this required or just preferred? I'd prefer to keep string types as flexible (or "future-proof") as possible.

Comment (alexanderdean) [reply schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0]: It's not required, but in Redshift you will end up with the Igluctl default lengths for every string - so you are not gaining a lot of flexibility...

**Comment by SyedWasiHaider** [comment]
@alexanderdean I am able to generate the sql but not sure what command is needed for the JSON paths. Is that something that needs to be done manually?

File: schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0 [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		},
+		"successfullyClosed": {
+			"type": "boolean"
+		},
+		"jsonPayload": {
+			"type": "number"
+		},
+		"referralApplication": {
+			"type": "number"
```
Comment (SyedWasiHaider) [inline schemas/com.mparticle.snowplow/appstatetransition_event/jsonschema/1-0-0]: This should also be string. Will fix.

**Comment by alexanderdean** [comment]
It's like this:

```
igluctl static generate --with-json-paths
```

**Comment by SyedWasiHaider** [comment]
Ok I think we're almost there. I may have one more commit to address the following:

- I'm going to consult internally about the max length for our string attributes
- Attempt do testing on a real Snowplow mini instance using the guide here: https://github.com/snowplow/snowplow-mini/wiki/Quickstart-guide and steps

**Comment by SyedWasiHaider** [comment]
@alexanderdean 

- With regards to the max string length, I think we're okay with the defaults for now. 

- The only question I have is what happens if you exceed the max string length? The Snowplow Mini instance seems to accept pretty much anything. For instance, if I send in a string field in any of our JSON with 10000 characters (well above the default 4096), it does show up in the "good" type in Kibana. So the question is: would the specific field basically be truncated or will the entire data be dropped?

**Comment by alexanderdean** [comment]
Hi @SyedWasiHaider - if you don't specify maxLengths on the strings in the JSON Schema, then the columns will just be truncated in Redshift.

**Comment by SyedWasiHaider** [comment]
Hi @alexanderdean,

We've done the testing on our side using the Snowplow mini instance and everything seems good. At this point, I think we're good with the schema definitions. We're also fine for now with the max string length and truncation behavior. Please give this a final pass and approve if everything looks good.

Thanks

**Comment by SyedWasiHaider** [comment]
Thanks so much Alex. When do you think your next release will be?

**Comment by SyedWasiHaider** [comment]
@alexanderdean Can the commits be in the same branch or do I need to create a new branch for each commit? I assume the former.

**Comment by SyedWasiHaider** [comment]
@alexanderdean Please let me know if anything looks incorrect here. Thanks.

**Comment by alexanderdean** [comment]
Many thanks @SyedWasiHaider, this looks great. Apologies for the busywork. We should get this released next week.

**Comment by alexanderdean** [comment]
Cherry-picked into https://github.com/snowplow/iglu-central/pull/648, closing

**Pull request closed without merging**